README.md
WebPaige Base

v2.3.2

CHANGELOG.md
# WebPaigeBase Change Log
----------------------------------------------------------------------------------------------------------------------------------------


## Installations
	- 100T	test environment
	- 110K	knrm live
	- 120G	gvrb
	- 130B	isparaat


## Versions

  2.3.4

	2.3.3
	  * Deployed for 100T, 110K, 120G and 130B 26-7-2013
		* Improved translations (#33) 12-7-2013
		* Implement the moving of time-slots starting in the past and ending in the future by slicing them 10-7-2013
		* Enabling changing of group member passwords for planners 9-7-2013
		* Multiple IE bugs fix for library loading order issue 9-7-2013

	2.3.2
		* Deployed for 100T, 110K, 120G and 130B 8-7-2013
		* Applying alphabetical sorting on receivers list when composing a message 8-7-2013
		* Internationalization updates with a focus on the Dutch language 8-7-2013
		* Major performance improvement when adding, modifying or deleting new timeslots (back-end improvement) 8-7-2013
		* Addressed a bug in the functionality of changing the availability of another team member 8-7-2013

	2.3.1
		* Deployed for 100T, 110K, 120G and 130B 3-7-2013
		* Hotfix for bug in changing view (e.g. jump back to week-view) after mutating planboard (#26/#3) 3-7-2013
		* Performance issues: Grouped calls for member timelines (#25) 28-6-2013
		* Performance issues: Live instance (#25) 28-6-2013
		* Background syncing for P2000 alerts (#32) 2-7-2013
		*	Extreme mutations in the past has been reviewed and blocked (#31) 2-7-2013
		* Show slotform feelds already filled with time and date when opening (#2) 1-7-2013
		* Adjust slotform times when mutation planboard items (#37) 1-7-2013
		* Auto-refresh kicks in  while mutating (#28) 1-7-2013
		* Remove slotform after planboard mutation (#36) 1-7-2013
		* Clicking today shows yesterday (#35) 1-7-2013
		* Set standard agenda view to today +6days (#30) 1-7-2013

	2.3.0
		* Deployed for 100T, 110K, 120G and 130B 27-5-2013
		* Always change a timeslot through slot form in planboard 27-6-2013
		* Planboard auto refresher is being blocked while working on planboard 27-6-2013
		* Percentage counter for members timeline in status bar 5-6-2013
		* Live 120G 5-6-2013
		* Help & Support 4-6-2013
		* Download mobile app functionality 4-6-2013
		* Deployed for 100T, 110K, 120G and 130B 16-5-2013

	2.2.0
		* Hotfix for not clearing cache for users 20-6-2013
		* Hotfix for stability issues in profile timeline 15-5-2013
		* Hotfix for not displaying user information on profile editing form 15-5-2013
		* Background planboard sync (60 seconds interval) 13-5-2013
		* Message job scheduler 13-5-2013
		* IE 8/9/10 planboard stabilization 13-5-2013
		* Hotfix for not going further than 6th page in pagination in messages 4-5-2013
		* Deployed for 100T, 110K, 120G and 130B 3-5-2013

	2.1.0
		* IE9 & IE10 support, IE8 (temporarily) and less are blocked 3-5-2013
		* Google Analytics impelemented 3-5-2013
		* Pagination in messages center 2-5-2013
		* Instruct user with email adres placeholder when remembering account info at login 29-4-2013
		* Conflict error message by registering same username 29-4-2013
		* Fixed not displaying alarm messages in dashboard 26-4-2013
		* Fixed centering styling issue for login button in login screen for IE8 26-4-2013
		* Language corrections in planboard statistics 19-4-2013
		* Hotfix for not displaying correct weeks by previous/future weeks in profile timeline 19-4-2013

	2.0.2
		* Deployed for 110K 17-4-2013
		* Core modules seperated 17-4-2013
		* Background syncing for every minute and displaying webkit desktop notification for new messages (still in development, temporarily disabled) 16-4-2013
		* By adding a new slot, slot form now opens with time of now and 1 hour later 15-4-2013
		* Full screen ability for webkit browsers 15-4-2013
		* Compressed third party libraries 15-4-2013
		* AMD support added for loading third-party-libraries (performance tweak) 15-4-2013
		* Major app loading performanace tweaks with reducing requests and minified and compiled to one version of app libraries 13-4-2013
		* Meridian time (hour) issue with selecting time range with mouse either by shift or ctrl keys is resolved 12-4-2013
		* Tablet issues has been solved 12-4-2013
		* Corrected a localization issue with 'no messages' in the box 11-4-2013
		* Tablet datetime selectors for planboard and profile timeline 11-4-2013
		* Removed footer for mobile and tablets and changed it to portable preloader 10-4-2013
		* Main navigation menu more friendly with apple or android tablets or mobiles 10-4-2013
		* First group to render setting functionality 10-4-2013
		* Default wish value setting for groups 9-4-2013
		* Dashboard 2.0 8-4-2013

	2.0.1
		* Fixed rendering issues with pie chart of totals in statistics in plan 5-4-2013
		* Dynamic listing for legenda based on available states of that particular installation 4-4-2013
		* Hot fix for not displaying date time conversions in message center 4-4-2013
		* Case-insesitive login 4-4-2013
		* Quick css fix for IE. (Not showing timeslots) 3-4-2013
		* CSS fix for tabs left border. (Bug with dynamic height sizing when there is no content in tab is still open!!) 3-4-2013
	
	2.0.0
		* Deployed for 100T, 110K, 120G, 130B 3-4-2013


## Changes in libraries
	timepicker.js
		* showMeridian property in libs/bootstrap-timepicker/timepicker.js changed to false for displayin 24 hours clock.

war/src/plugins.js
/**
 * Avoid `console` errors in browsers that lack a console
 */
(function ()
{
  var method,
      noop = function() {},
      methods = [
        'assert', 'clear', 'count', 'debug', 'dir', 'dirxml', 'error',
        'exception', 'group', 'groupCollapsed', 'groupEnd', 'info', 'log',
        'markTimeline', 'profile', 'profileEnd', 'table', 'time', 'timeEnd',
        'timeStamp', 'trace', 'warn'
      ],
      length = methods.length,
      console = (window.console = window.console || {});

  while (length--)
  {
    method = methods[length];

    if (!console[method]) console[method] = noop;
  }
}());;/**
 * jQuery $.browser plugin
 * Depreciated since 1.9
 *
 * https://github.com/gabceb/jquery-browser-plugin
 */
(function (jQuery, window, undefined) 
{
  "use strict";

  var matched, browser;

  jQuery.uaMatch = function(ua) {
    ua = ua.toLowerCase();

    var match = /(chrome)[ \/]([\w.]+)/.exec(ua) ||
      /(webkit)[ \/]([\w.]+)/.exec(ua) ||
      /(opera)(?:.*version|)[ \/]([\w.]+)/.exec(ua) ||
      /(msie) ([\w.]+)/.exec(ua) || ua.indexOf("compatible") < 0 && /(mozilla)(?:.*? rv:([\w.]+)|)/.exec(ua) || [];

    var platform_match = /(ipad)/.exec(ua) ||
      /(iphone)/.exec(ua) ||
      /(android)/.exec(ua) || [];

    return {
      browser: match[1] || "",
      version: match[2] || "0",
      platform: platform_match[0] || ""
    };
  };

  matched = jQuery.uaMatch(window.navigator.userAgent);
  browser = {};

  if (matched.browser) 
  {
    browser[matched.browser] = true;
    browser.version = matched.version;
  };

  if (matched.platform) 
  {
    browser[matched.platform] = true
  };

  // Chrome is Webkit, but Webkit is also Safari.
  if (browser.chrome) 
  {
    browser.webkit = true;
  } 
  else if (browser.webkit) 
  {
    browser.safari = true;
  };

  // check for mobile
  if ( /Android|webOS|iPhone|iPad|iPod|BlackBerry/i.test(window.navigator.userAgent) )
  {
    browser.mobile = true;

    browser.android = (/Android/i.test(window.navigator.userAgent)) ? true : false;

    browser.ios     = (/iPhone|iPad|iPod/i.test(window.navigator.userAgent)) ? true : false;
  }
  else
  {
    browser.mobile = false;
  };

  jQuery.browser = browser;

})(jQuery, window);




/**
 * Detect IE version for blocking IE6 and IE7
 */
if ($.browser.msie)
{
  var ver = $.browser.version || $.browser.version[0];

  if (ver == '6.0' || ver == '7.0') window.location = 'browsers.html';
}


;/**
 * jQuery $.os plugin
 *
 * Author: Cengiz Ulusoy
 */
(function (jQuery, window, undefined) 
{
  "use strict";

  /**
   * Init vars
   */
  var os = {},
      navOS = window.navigator.appVersion;

  /**
   * Os tests
   */
  if (navOS.indexOf("Win") != -1)   os.windows = true;
  if (navOS.indexOf("Mac") != -1)   os.mac = true;
  if (navOS.indexOf("X11") != -1)   os.unix = true;
  if (navOS.indexOf("Linux") != -1) os.linux = true;

  /**
   * Pass to jQuery
   */
  jQuery.os = os;

})(jQuery, window);;/*!
* basket.js
* v0.3.0 - 2012-12-28
* http://addyosmani.github.com/basket.js
* (c) Addy Osmani; MIT License
* Created by: Addy Osmani, Sindre Sorhus, AndrÃ©e Hansson
* Contributors: Ironsjp, Mathias Bynens, Rick Waldron, Felipe Morais
* Uses rsvp.js, https://github.com/tildeio/rsvp.js
*/
if ('localStorage' in window && window['localStorage'] !== null)
{
  (function( window, document ) {
    'use strict';

    // Monkey-patching an "all" method onto RSVP
    // Returns a promise that will be fulfilled when the array of promises passed in are all
    // fulfilled
    RSVP.all = function( promises ) {
      var i, results = [];
      var allPromise = new RSVP.Promise();
      var remaining = promises.length;

      var resolver = function( index ) {
        return function( value ) {
          resolve( index, value );
        };
      };
      var resolve = function( index, value ) {
        results[ index ] = value;
        if ( --remaining === 0 ) {
          allPromise.resolve( results );
        }
      };
      var reject = function( error ) {
        allPromise.reject( error );
      };

      for ( i = 0; i < remaining; i++ ) {
        promises[ i ].then( resolver( i ), reject );
      }

      return allPromise;
    };

    var head = document.head || document.getElementsByTagName('head')[0];
    var storagePrefix = 'WebPaige.';
    var defaultExpiration = 5000;

    var addLocalStorage = function( key, storeObj ) {
      try {
        localStorage.setItem( storagePrefix + key, JSON.stringify( storeObj ) );
        return true;
      } catch( e ) {
        if ( e.name.toUpperCase().indexOf('QUOTA') >= 0 ) {
          var item;
          var tempScripts = [];

          for ( item in localStorage ) {
            if ( item.indexOf( storagePrefix ) === 0 ) {
              tempScripts.push( JSON.parse( localStorage[ item ] ) );
            }
          }

          if ( tempScripts.length ) {
            tempScripts.sort(function( a, b ) {
              return a.stamp - b.stamp;
            });

            basket.remove( tempScripts[ 0 ].key );

            return addLocalStorage( key, storeObj );

          } else {
            // no files to remove. Larger than available quota
            return;
          }

        } else {
          // some other error
          return;
        }
      }

    };

    var getUrl = function( url ) {
      var xhr = new XMLHttpRequest();
      var promise = new RSVP.Promise();
      xhr.open( 'GET', url );

      xhr.onreadystatechange = function() {
        if ( xhr.readyState === 4 ) {
          if( xhr.status === 200 ) {
            promise.resolve( xhr.responseText );
          } else {
            promise.reject( new Error( xhr.statusText ) );
          }
        }
      };

      xhr.send();

      return promise;
    };

    var saveUrl = function( obj ) {
      return getUrl( obj.url ).then( function( text ) {
        var storeObj = wrapStoreData( obj, text );

        addLocalStorage( obj.key , storeObj );

        return text;
      });
    };

    var injectScript = function( text ) {
      var script = document.createElement('script');
      script.defer = true;
      // Have to use .text, since we support IE8,
      // which won't allow appending to a script
      script.text = text;
      head.appendChild( script );
    };

    var wrapStoreData = function( obj, data ) {
      var now = +new Date();
      obj.data = data;
      obj.stamp = now;
      obj.expire = now + ( ( obj.expire || defaultExpiration ) * 60 * 60 * 1000 );

      return obj;
    };

    var handleStackObject = function( obj ) {
      var source, promise;

      if ( !obj.url ) {
        return;
      }

      obj.key =  ( obj.key || obj.url );
      source = basket.get( obj.key );

      obj.execute = obj.execute !== false;

      if ( !source || source.expire - +new Date() < 0  || obj.unique !== source.unique ) {
        if ( obj.unique ) {
          // set parameter to prevent browser cache
          obj.url += ( ( obj.url.indexOf('?') > 0 ) ? '&' : '?' ) + 'basket-unique=' + obj.unique;
        }
        promise = saveUrl( obj );
      } else {
        promise = new RSVP.Promise();
        promise.resolve( source.data );
      }

      if( obj.execute ) {
        return promise.then( injectScript );
      } else {
        return promise;
      }
    };

    window.basket = {
      require: function() {
        var i, l, promises = [];

        for ( i = 0, l = arguments.length; i < l; i++ ) {
          promises.push( handleStackObject( arguments[ i ] ) );
        }

        return RSVP.all( promises );
      },

      remove: function( key ) {
        localStorage.removeItem( storagePrefix + key );
        return this;
      },

      get: function( key ) {
        var item = localStorage.getItem( storagePrefix + key );
        try {
          return JSON.parse( item || 'false' );
        } catch( e ) {
          return false;
        }
      },

      clear: function( expired ) {
        var item, key;
        var now = +new Date();

        for ( item in localStorage ) {
          key = item.split( storagePrefix )[ 1 ];
          if ( key && ( !expired || this.get( key ).expire <= now ) ) {
            this.remove( key );
          }
        }

        return this;
      }
    };

    // delete expired keys
    basket.clear( true );

  })( this, document );
};/**
 * Screenfull - Sindresorhus
 * 
 * https://github.com/sindresorhus/screenfull.js
 */
(function(window, document) {
  'use strict';

  var keyboardAllowed = typeof Element !== 'undefined' && 'ALLOW_KEYBOARD_INPUT' in Element, // IE6 throws without typeof check

    fn = (function() {
      var val, valLength;
      var fnMap = [
        [
          'requestFullscreen',
          'exitFullscreen',
          'fullscreenElement',
          'fullscreenEnabled',
          'fullscreenchange',
          'fullscreenerror'
        ],
        // new WebKit
        [
          'webkitRequestFullscreen',
          'webkitExitFullscreen',
          'webkitFullscreenElement',
          'webkitFullscreenEnabled',
          'webkitfullscreenchange',
          'webkitfullscreenerror'

        ],
        // old WebKit (Safari 5.1)
        [
          'webkitRequestFullScreen',
          'webkitCancelFullScreen',
          'webkitCurrentFullScreenElement',
          'webkitCancelFullScreen',
          'webkitfullscreenchange',
          'webkitfullscreenerror'

        ],
        [
          'mozRequestFullScreen',
          'mozCancelFullScreen',
          'mozFullScreenElement',
          'mozFullScreenEnabled',
          'mozfullscreenchange',
          'mozfullscreenerror'
        ]
      ];
      var i = 0;
      var l = fnMap.length;
      var ret = {};

      for (; i < l; i++) {
        val = fnMap[i];
        if (val && val[1] in document) {
          for (i = 0, valLength = val.length; i < valLength; i++) {
            ret[fnMap[0][i]] = val[i];
          }
          return ret;
        }
      }
      return false;
    })(),

    screenfull = {
      request: function(elem) {

        var request = fn.requestFullscreen;

        elem = elem || document.documentElement;

        // Work around Safari 5.1 bug: reports support for
        // keyboard in fullscreen even though it doesn't.
        // Browser sniffing, since the alternative with
        // setTimeout is even worse
        if (/5\.1[\.\d]* Safari/.test(navigator.userAgent)) {
          elem[request]();
        } else {
          elem[request](keyboardAllowed && Element.ALLOW_KEYBOARD_INPUT);
        }
      },
      exit: function() {
        document[fn.exitFullscreen]();
      },
      toggle: function( elem ) {
        if (this.isFullscreen) {
          this.exit();
        } else {
          this.request(elem);
        }
      },
      onchange: function() {},
      onerror: function() {}
    };

  if (!fn) {
    return window.screenfull = false;
  }

  Object.defineProperties(screenfull, {
    isFullscreen: {
      get: function() {
        return !!document[fn.fullscreenElement];
      }
    },
    element: {
      enumerable: true,
      get: function() {
        return document[fn.fullscreenElement];
      }
    },
    enabled: {
      enumerable: true,
      get: function() {
        // Coerce to boolean in case of old WebKit
        return !!document[fn.fullscreenEnabled];
      }
    }
  });

  document.addEventListener(fn.fullscreenchange, function(e) {
    screenfull.onchange.call(screenfull, e);
  });

  document.addEventListener(fn.fullscreenerror, function(e) {
    screenfull.onerror.call(screenfull, e);
  });

  window.screenfull = screenfull;

})(window, document);